internal/core/taskqueue.go

```diff
@@ -8,82 +8,60 @@ import (
 	"sync"
 	"time"
 
+	"github.com/SwissOpenEM/Ingestor/internal/task"
+	"github.com/elliotchance/orderedmap/v2"
 	"github.com/google/uuid"
 )
 
 type TaskQueue struct {
-	datasetSourceFolders sync.Map           // Datastructure to store all the upload requests
-	inputChannel         chan IngestionTask // Requests to upload data are put into this channel
-	resultChannel        chan TaskResult    // The result of the upload is put into this channel
-	AppContext           context.Context
-	Config               Config
-	Notifier             ProgressNotifier
-}
-
-type TransferMethods int
-
-const (
-	TransferS3 TransferMethods = iota + 1
-	TransferGlobus
-)
-
-type TransferOptions struct {
-	S3_endpoint string
-	S3_Bucket   string
-	Md5checksum bool
-}
-
-type TaskTransferConfig struct {
-	S3TransferConfig
-	GlobusTransferConfig
-}
-
-type IngestionTask struct {
-	// DatasetFolderId   uuid.UUID
-	DatasetFolder
-	TransferMethod TransferMethods
-	Cancel         context.CancelFunc
-}
-
-type TaskResult struct {
-	Elapsed_seconds int
-	Dataset_PID     string
-	Error           error
+	taskListLock       sync.RWMutex                                          // locking mechanism for uploadIds and datasetUploadTasks
```

Comment:
I need to have an ordered map or a way to store the order in which elements got added to the map. This is because I need to return a list of elements on one of the endpoints with pagination. I could've just added a separate slice next to the sync.Map, but that would still require adding a mutex lock system, so I just opted to use an ordered map directly

Reply:
ok, fair enough, thanks.